File: backend/app/schemas/workspace.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.asset import AssetResponse
from app.schemas.business import BusinessResponse
from app.schemas.campaign import CampaignResponse
from app.schemas.concept import CreativeConceptResponse
from app.schemas.final_video import FinalVideoResponse
from app.schemas.scene import SceneResponse
from app.schemas.storyboard import StoryboardResponse
from app.schemas.strategy import CampaignStrategyResponse


class SceneDetailResponse(SceneResponse):
    assets: List[AssetResponse] = Field(default_factory=list)
    selected_asset: Optional[AssetResponse] = None


class CampaignProgressResponse(BaseModel):
    brief_completed: bool
    strategy_completed: bool
    concept_selected: bool
    storyboard_ready: bool
    scenes_generated: bool
    scenes_generated_count: int
    scenes_total_count: int
    quality_evaluated: bool
    final_rendered: bool
    current_stage: str = Field(description="Stage: BRIEF, STRATEGY, CONCEPTS, STORYBOARD, GENERATION, EVALUATION, RENDER, COMPLETED")
    progress_percentage: int = Field(description="Overall workflow completion percentage from 0 to 100")


class CampaignWorkspaceResponse(BaseModel):
    campaign: CampaignResponse
    business: Optional[BusinessResponse] = None
    strategy: Optional[CampaignStrategyResponse] = None
    concepts: List[CreativeConceptResponse] = Field(default_factory=list)
    selected_concept: Optional[CreativeConceptResponse] = None
    storyboard: Optional[StoryboardResponse] = None
    scenes: List[SceneDetailResponse] = Field(default_factory=list)
    final_videos: List[FinalVideoResponse] = Field(default_factory=list)
    progress: CampaignProgressResponse

    model_config = ConfigDict(from_attributes=True)


class SceneCreateRequest(BaseModel):
    sequence_number: Optional[int] = Field(default=None, description="Optional target position; appends to end if omitted")
    shot_type: str = Field(default="Medium Tracking Shot")
    camera_movement: str = Field(default="Dolly In")
    visual_prompt: str
    audio_narration: str = ""
    duration_seconds: float = 4.0
    lighting_atmosphere: str = "Natural atmospheric commercial lighting"


class SceneReorderRequest(BaseModel):
    scene_ids_in_order: List[UUID] = Field(description="Ordered list of scene UUIDs defining new sequence order")

File: frontend/src/features/campaigns/types/workspace.ts
export interface BusinessData {
  id: string;
  name: string;
  industry: string;
  description?: string;
  website_url?: string;
  target_audience?: string;
  tone_of_voice?: string;
  brand_colors?: string;
  brand_guidelines?: string;
}

export interface CampaignData {
  id: string;
  business_id: string;
  name: string;
  goal: string;
  target_audience?: string;
  budget?: number;
  duration_seconds: number;
  aspect_ratio: '16:9' | '9:16' | '1:1' | '4:5' | string;
  status: 'draft' | 'strategy_generated' | 'storyboard_ready' | 'rendering' | 'completed' | 'failed';
  created_at: string;
  updated_at: string;
}

export interface StrategyData {
  id: string;
  campaign_id: string;
  target_audience_breakdown?: Record<string, unknown> | string;
  value_propositions?: string[] | string;
  emotional_triggers?: string[] | string;
  messaging_pillars?: string[] | string;
}

export interface CreativeConceptData {
  id: string;
  campaign_id: string;
  title: string;
  angle_type: string;
  hook: string;
  visual_style: string;
  narrative_arc: string;
  target_emotion: string;
  is_selected: boolean;
}

export interface AssetData {
  id: string;
  scene_id?: string;
  asset_type: string;
  storage_url: string;
  version_number: number;
  status: string;
}

export interface SceneDetailData {
  id: string;
  storyboard_id: string;
  sequence_number: number;
  shot_type: string;
  camera_movement: string;
  visual_prompt: string;
  audio_narration: string;
  duration_seconds: number;
  lighting_atmosphere: string;
  assets: AssetData[];
  selected_asset?: AssetData;
}

export interface FinalVideoData {
  id: string;
  timeline_id: string;
  resolution: string;
  duration_seconds: number;
  video_url: string;
  aspect_ratio: string;
  status: string;
  created_at: string;
}

export interface CampaignProgressData {
  brief_completed: boolean;
  strategy_completed: boolean;
  concept_selected: boolean;
  storyboard_ready: boolean;
  scenes_generated: boolean;
  scenes_generated_count: number;
  scenes_total_count: number;
  quality_evaluated: boolean;
  final_rendered: boolean;
  current_stage: 'BRIEF' | 'STRATEGY' | 'CONCEPTS' | 'STORYBOARD' | 'GENERATION' | 'EVALUATION' | 'RENDER' | 'COMPLETED';
  progress_percentage: number;
}

export interface CampaignWorkspaceData {
  campaign: CampaignData;
  business?: BusinessData;
  strategy?: StrategyData;
  concepts: CreativeConceptData[];
  selected_concept?: CreativeConceptData;
  storyboard?: {
    id: string;
    campaign_id: string;
    creative_bible?: Record<string, unknown>;
  };
  scenes: SceneDetailData[];
  final_videos: FinalVideoData[];
  progress: CampaignProgressData;
}
